Add engineering onboarding doc to handbook, small organizational changes to onboarding section of handbook (#198)
mattermost/mattermost-handbook: contributors/onboarding/staff-onboarding-guide/engineer-onboarding.md

> # Mattermost Engineer Onboarding Timeline & Expectations

Reviewer: Great to see this surfaced in the handbook! Looks good to me.

One note is that I'm not sure this can fully /replace/ https://docs.google.com/document/d/14jsUJgGO4j5f4B6bPVCIC1y9yXy43rmAFI0F5MJsuE0/edit, since I've copied the original document for each new staff member I manager and use it as a sort of 1:1 checklist to ensure alignment. It's meant to be a sort of living document in their context as well, with Google Docs being a perfect medium for sharing comments and such.

Reviewer 2: Thanks for raising this. My opinion is that this page is more of an outline for candidates/new staff to give an overview of expectations and process. I don't see a problem with using a Google doc like yours unless it deviates substantially from this content. I would expect each team/manager to have items that aren't mentioned in this document so a bit of customization shouldn't cause confusion. 

Totally my opinion though, @jasonblais needs to weigh in as well.

Author: @lieut-data I'm wondering if the same steps could be taken from the handbook -- which is to create a fresh Google Doc and copy-paste the content from the handbook. 

It does add an extra step into that workflow, but I worry the handbook page will become outdated if the Google Doc is continued to be used as the source of truth.

Author: (And if the Google Doc is continued to be used, then perhaps we should just link to that doc from the handbook instead)

Reviewer: I'm fine with the copy/paste proposal. @jwilander, I know you wrote the original version of this very helpful article. Any thoughts?

Reviewer 3: I like having the source of truth copy in the handbook 👍 I'm good with copy/pasting it into a google doc as needed

This is the general timeline and expectations for new engineers starting at Mattermost. It is meant to guide new engineers, help them focus on the right tasks, and set expectations to help them be successful at Mattermost. It is not a hard timeline in any sense and circumstances may shorten, lengthen or re-order the timeline.

## Week 1: Focus on environment setup and introductions

* Set up equipment and environment (e.g. complete the day 1 checklist)
  * Get laptop, set up developer environment
  * Clone enterprise repo
  * Log in to all accounts (Gmail, GitHub, JIRA, OneLogin, Community, Community Daily)
  * Expense equipment
  * HR set up (benefits, health insurance, etc.)
* Meet the team

>   * Meet with a different team member for 30min each day

Reviewer 2: ```suggestion
  * Meet with a different team member for 30 minutes each day
```

* Meet with your dev lead

>   * Meet for at least 10-15 minutes everyday

Reviewer 2: ```suggestion
  * Meet for at least 10-15 minutes each day
```

* Attend weekly team/company meetings to get the feeling for them
  * Sprint Planning
  * Company All Hands
  * Developer Meeting
  * R&D Meeting
  * Developer Hangouts
  * Video Game Hour
* Join important channels
  * https://community.mattermost.com/core/channels/developers
  * https://community.mattermost.com/core/channels/developers-meeting
  * https://community.mattermost.com/core/channels/thank-you
  * https://community.mattermost.com/private-core/channels/announcements
  * https://community.mattermost.com/private-core/channels/confidential-bugs

>   * https://community.mattermost.com/private-core/channels/off-topic

Author: Changed link from community-release to community

* Read and understand the core values
  * https://handbook.mattermost.com/company/about-mattermost#leadership-principles
* Review video calling etiquette
  * http://hulen.com/video-conf-peevs/
* Work on 1-3 small tickets to get used to dev processes
* Introduction at the R&D Meeting

## Week 2: Focus on digesting information dump

* Continue getting used to processes and workflows
  * Sit in on some spec/design/technical meetings, your dev lead should send you invites to these meetings
  * Do some code reviews, your dev lead should assign you to some
* Introduction meetings

>   * Meet with a different team member for 30min each day

Reviewer 2: ```suggestion
  * Meet with a different team member for 30 minutes each day
```

* Mentor for the week is a senior teammate

>   * Meet for at least 10-15 minutes everyday

Reviewer 2: ```suggestion
  * Meet for at least 10-15 minutes each day
```

* Meet with dev lead
  * Meet 2-3 times in the week
  * Go over core values with dev lead
* Work on 1-3 small tickets

> * Introduction at the [Customer Obsession Meeting](https://handbook.mattermost.com/operations/operations/company-cadence#customer-obsession-meeting-aka-com)

Author: Renamed meeting to "Customer Obsession Meeting" from "Company All Hands meeting", linked to handbook entry to learn more about the meeting

## Weeks 3-4: Focus on solidifying role in the team

* Start taking a more active role as a member of your dev team
* Begin to participate in some spec/design/technical meetings
  * Provide feedback/ask questions, have an active role in the meetings
* Mentor for each week is a different teammate

>   * Meet for at least 10-15 minutes everyday

Reviewer 2: ```suggestion
  * Meet for at least 10-15 minutes each day
```

* Meet with dev lead
  * Meet once a week and start regular weekly 1-1s
* Work on some medium size tickets

## Weeks 5-8 (Month 2): Work on your first project as dev owner

> * Begin work on your first large ish task and be the dev owner of it

Reviewer 2: ```suggestion
* Begin work on your first relatively large task (40 mana) and be the dev owner of it
```
Can we qualify "relatively large"? Maybe we should add the mana expectation?

Author: ^ @lieut-data this would be for you (or another dev lead) to confirm

Reviewer: I'm fine with `relative large` task, and might define that as 40 mana.

* Act as the dev owner in the spec/design/technical meetings
* Write up a brief technical specification if necessary
* Work on small/medium tickets each sprint alongside the larger task
* It is expected that your first project takes longer to complete than usual

## Weeks 9-11 (Month 3): Work on more and/or larger projects as dev owner

* Take what you learned being dev owner on your first project and improve upon it
* Act as dev owner and write technical specs for more and larger projects
* Work on small/medium tickets to fill in gaps
* Complete deliverables in a more timely fashion
* Attend a "Tech Moonshots" developer meeting

## Week 12: Informal performance evaluation

* Dev lead will give you your performance evaluation
* A self survey will be sent to you

> * Surveys will be sent out to your peers

Reviewer 2: ```suggestion
* Surveys will be sent out to your peers and you are:
```

  * Evaluated on what is being done well, what could be improved
  * Evaluated on core values

> * Goal here is to let you know how you’re performing and to help you improve and be a better engineer

Reviewer 2: ```suggestion
* The goal is to let you know how you’re performing and to help you improve and be a better engineer
```

## Weeks 13-16 (Month 4): Act on your performance evaluation and focus on community

* Continue to improve on fulfilling engineering responsibilities and completing quality deliverables in a timely fashion
* Begin acting on any feedback given during the performance evaluation
* Start taking a more active role in the open source community
  * Proactively answer questions in channels

>   * Create help wanted tickets and mentor contributors working on them

Reviewer 2: ```suggestion
  * Create Help Wanted tickets and mentor contributors working on them
```

  * Find a community buddy (a contributor you interact with fairly often)
* Start helping to answer customer support questions

## Weeks 17-20 (Month 5): Become an authority

* Start to become an authority on some part of the product/codebase
  * Begin thinking about what part of the code you want to “own”, what do you want to be known for at Mattermost
  * Could be something you built on a past project
  * Be the “go to” person for that area
  * Tends to happen naturally if you take true ownership of the projects you’ve worked on

> * Create some help wanted tickets/campaigns to get the community involved in your area of authority

Reviewer 2: ```suggestion
* Create some Help Wanted tickets/campaigns to get the community involved in your area of authority
```
